% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\usepackage[utf8]{inputenc}
\DeclareUnicodeCharacter{00A0}{\nobreakspace}
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage{longtable}
\usepackage{sphinx}
\usepackage{multirow}

\addto\captionsenglish{\renewcommand{\figurename}{Fig. }}
\addto\captionsenglish{\renewcommand{\tablename}{Table }}
\floatname{literal-block}{Listing }



\title{pyCGA Documentation}
\date{September 04, 2015}
\release{0.1.0}
\author{Antonio Rueda, Mathew Parker, Razvan Sultana}
\newcommand{\sphinxlogo}{}
\renewcommand{\releasename}{Release}
\makeindex

\makeatletter
\def\PYG@reset{\let\PYG@it=\relax \let\PYG@bf=\relax%
    \let\PYG@ul=\relax \let\PYG@tc=\relax%
    \let\PYG@bc=\relax \let\PYG@ff=\relax}
\def\PYG@tok#1{\csname PYG@tok@#1\endcsname}
\def\PYG@toks#1+{\ifx\relax#1\empty\else%
    \PYG@tok{#1}\expandafter\PYG@toks\fi}
\def\PYG@do#1{\PYG@bc{\PYG@tc{\PYG@ul{%
    \PYG@it{\PYG@bf{\PYG@ff{#1}}}}}}}
\def\PYG#1#2{\PYG@reset\PYG@toks#1+\relax+\PYG@do{#2}}

\expandafter\def\csname PYG@tok@gd\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gu\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@gt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PYG@tok@gs\endcsname{\let\PYG@bf=\textbf}
\expandafter\def\csname PYG@tok@gr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@cm\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@vg\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@m\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@mh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@cs\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\colorbox[rgb]{1.00,0.94,0.94}{\strut ##1}}}
\expandafter\def\csname PYG@tok@ge\endcsname{\let\PYG@it=\textit}
\expandafter\def\csname PYG@tok@vc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@il\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@go\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.20,0.20,0.20}{##1}}}
\expandafter\def\csname PYG@tok@cp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@gi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gh\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@ni\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.84,0.33,0.22}{##1}}}
\expandafter\def\csname PYG@tok@nl\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.13,0.44}{##1}}}
\expandafter\def\csname PYG@tok@nn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@no\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.38,0.68,0.84}{##1}}}
\expandafter\def\csname PYG@tok@na\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@nb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@nd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.33,0.33,0.33}{##1}}}
\expandafter\def\csname PYG@tok@ne\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.49}{##1}}}
\expandafter\def\csname PYG@tok@si\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.44,0.63,0.82}{##1}}}
\expandafter\def\csname PYG@tok@s2\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@vi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@nt\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.45}{##1}}}
\expandafter\def\csname PYG@tok@nv\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@s1\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@gp\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@sh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@ow\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@sx\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@bp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c1\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@kc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@mf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@err\endcsname{\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PYG@tok@mb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@ss\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.32,0.47,0.09}{##1}}}
\expandafter\def\csname PYG@tok@sr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.14,0.33,0.53}{##1}}}
\expandafter\def\csname PYG@tok@mo\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@kd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@mi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@kn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@o\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PYG@tok@kr\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@s\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@kp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@w\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PYG@tok@kt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.56,0.13,0.00}{##1}}}
\expandafter\def\csname PYG@tok@sc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@k\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@se\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sd\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}

\def\PYGZbs{\char`\\}
\def\PYGZus{\char`\_}
\def\PYGZob{\char`\{}
\def\PYGZcb{\char`\}}
\def\PYGZca{\char`\^}
\def\PYGZam{\char`\&}
\def\PYGZlt{\char`\<}
\def\PYGZgt{\char`\>}
\def\PYGZsh{\char`\#}
\def\PYGZpc{\char`\%}
\def\PYGZdl{\char`\$}
\def\PYGZhy{\char`\-}
\def\PYGZsq{\char`\'}
\def\PYGZdq{\char`\"}
\def\PYGZti{\char`\~}
% for compatibility with earlier versions
\def\PYGZat{@}
\def\PYGZlb{[}
\def\PYGZrb{]}
\makeatother

\renewcommand\PYGZsq{\textquotesingle}

\begin{document}

\maketitle
\tableofcontents
\phantomsection\label{documentation::doc}


\href{https://github.com/opencb/opencga/wiki}{OpenCGA}  is an open-source project that aims to provide a Big Data storage engine and analysis
framework for genomic scale data analysis of hundreds of terabytes or even petabytes. For users,
its main features will include uploading and downloading files to a repository, storing their
information in a generic way (non-dependant of the original file-format) and retrieving this information
efficiently.

The aim of this python package is provide an interface using the OpenCGA web services from python, and make
easy and comprehensive the use of OpenCGA. And also provide a CLI usable from

The code is open source, and \href{https://github.com/opencb/}{available on github}.

The main documentation for the project is organized into a couple sections:
\begin{itemize}
\item {} 
{\hyperref[documentation:getting-started]{\emph{\DUspan{}{Getting Started}}}}

\item {} 
{\hyperref[documentation:api-docs]{\emph{\DUspan{}{API Documentation}}}}

\item {} 
{\hyperref[documentation:cli-docs]{\emph{\DUspan{}{CLI Documentation}}}}

\item {} 
{\hyperref[documentation:dev-docs]{\emph{\DUspan{}{Developer Documentation}}}}

\end{itemize}


\chapter{Getting Started}
\label{documentation:welcome-to-pycga}\label{documentation:getting-started}\label{documentation:id1}

\section{Installation}
\label{installation:installation}\label{installation::doc}
Here is a step by step guide on how to install pyCGA. It will get you the library ready to be load from a python script and all the scripts we provide installed in your system.

First, obtain Python if you do not already have it. Then you will ned to install \href{https://pip.pypa.io/en/latest/installing.html}{pip}

Once you have these you will be able to install the requirements


\subsection{Requirements}
\label{installation:requirements}
\begin{tabulary}{\linewidth}{|L|L|}
\hline
\textsf{\relax 
Packages
} & \textsf{\relax 
Version
}\\
\hline
requests
 & 
2.7
\\
\hline\end{tabulary}


Although all the packages, The dependencies are specified in the file pyCGA/pip\_requirements.txt
You can install them using:

\begin{Verbatim}[commandchars=\\\{\}]
cd pyCGA
pip install \PYGZhy{}r pip\PYGZus{}requirements.txt
\end{Verbatim}


\subsection{Install}
\label{installation:install}
Next step to complete the installation is execute the file setup.py, you can use:

\begin{Verbatim}[commandchars=\\\{\}]
sudo python setup.py install
\end{Verbatim}

or:

\begin{Verbatim}[commandchars=\\\{\}]
sudo pip install .
\end{Verbatim}

\begin{notice}{note}{Note:}
If you are using pip and you would like to reinstall the packages you should use:

\begin{Verbatim}[commandchars=\\\{\}]
sudo pip install . \PYGZhy{}\PYGZhy{}upgrade
\end{Verbatim}
\end{notice}


\section{Quick-start}
\label{quick_start:quick-start}\label{quick_start::doc}
inf


\section{Id Considerations in OpenCGA}
\label{ids:how-to-use-ids}\label{ids::doc}\label{ids:id-considerations-in-opencga}

\chapter{API Documentation}
\label{documentation:api-docs}\label{documentation:api-documentation}

\section{Web Services API}
\label{web_services:web-services-api}\label{web_services::doc}
inf


\section{How To Use API from python}
\label{queries_from_python:how-to-use-api-from-python}\label{queries_from_python::doc}
inf


\chapter{CLI Documentation}
\label{documentation:cli-documentation}\label{documentation:cli-docs}

\section{PyCGA}
\label{pyCGA:pycga}\label{pyCGA::doc}
inf


\section{pyCGAVariantFetcher}
\label{pyVariantFetcher:pycgavariantfetcher}\label{pyVariantFetcher::doc}

\subsection{CommandLine Options}
\label{pyVariantFetcher:commandline-options}

This program can be used to fetch variant easily from openCGA

\begin{alltt}
usage: pyCGAVariantFetcher {[}-h{]} --host OpenCGA Host --studyID study Id {[}--ids ids{]} {[}--region region{]} {[}--chromosome chromosome{]} {[}--gene gene{]}
                           {[}--type type{]} {[}--reference reference{]} {[}--alternate alternate{]} {[}--files files{]} {[}--maf maf{]} {[}--mgf mgf{]}
                           {[}--missingAlleles missingAlleles{]} {[}--missingGenotypes missingGenotypes{]} {[}--annotationExists{]} {[}--annotationDoesNotExist{]}
                           {[}--genotype genotype{]} {[}--consequence\_type annot-ct{]} {[}--xref xref{]} {[}--biotype biotype{]} {[}--polyphen polyphen{]} {[}--sift sift{]}
                           {[}--conservation conservation{]} {[}--alternate\_frequency alternate\_frequency{]} {[}--reference\_frequency reference\_frequency{]}
                           {[}--limit reference\_frequency{]} {[}--skip reference\_frequency{]} {[}--sort{]} {[}--group\_by reference\_frequency{]}
                           {[}--unknownGenotype unknownGenotype{]} {[}--returnedSamples returnedSamples{]} {[}--returnedFiles returnedFiles{]}
                           {[}--batchSize batchSize{]} {[}--outputType output{]}
                           Session Id
\end{alltt}
\begin{description}
\item[{Positional arguments:}] \leavevmode\begin{optionlist}{3cm}
\item [sid]  
Required. To get this session ID yoy need to login using pyCGA
\end{optionlist}

\item[{Options:}] \leavevmode\begin{optionlist}{3cm}
\item [-{-}host]  
Required. This is the url to web service host. (i,e http://XX.X.XX.XXX:8080/)
\item [-{-}studyID]  
Required. The id of the study, this can be query usgin pyCGA
\item [-{-}ids]  
Optional. Select by variant ids (i,e dbSNP ids). Use comma to separate the ids
\item [-{-}region]  
Optional. Select by region (i,e chr:start-end). Use comma to separate the region
\item [-{-}chromosome]  
Optional. Select by chromosome. Use comma to separate the region
\item [-{-}gene]  
Optional. Select by gene name (It can be ENSMBL Ids or HGNCs symbols. Use comma to separate the gene names
\item [-{-}type]  
Optional. Select by type of variant  {[}SNV, MNV, INDEL, SV, CNV{]}. This filter should not be used alone due to the huge amount of results could be retrieved
\item [-{-}reference]  
Optional. Select by reference base(s)
\item [-{-}alternate]  
Optional. Select by alternate base(s)
\item [-{-}files]  
Optional. Select by files, only the variants found in these file will be selected. Please note these are the id files in the DB, use pyCGA to get the id files. Use comma to separate the file ids
\item [-{-}maf]  
Optional. Select by minor allele frequency. In this filter only the samples in the db are consider, not any external population. Syntax: {[}\textless{}\textbar{}\textgreater{}\textbar{}\textless{}=\textbar{}\textgreater{}={]}\{number\}, (e.g, \textgreater{}=0.05)
\item [-{-}mgf]  
Optional. Select by minor genotype frequency. In this filter only the samples in the db are consider, not any external population. Syntax: {[}\textless{}\textbar{}\textgreater{}\textbar{}\textless{}=\textbar{}\textgreater{}={]}\{number\}, (e.g, \textgreater{}=0.1)
\item [-{-}missingAlleles]  
Optional. Select by number of missing alleles in the whole sample set. In this filter only the samples in the db are consider, not any external population. Syntax: {[}\textless{}\textbar{}\textgreater{}\textbar{}\textless{}=\textbar{}\textgreater{}={]}\{number\}, (e.g, \textgreater{}=0.05)
\item [-{-}missingGenotypes]  
Optional. Select by number of missing genotypes in the whole sample set. In this filter only the samples in the db are consider, not any external population. Syntax: {[}\textless{}\textbar{}\textgreater{}\textbar{}\textless{}=\textbar{}\textgreater{}={]}\{number\}, (e.g, \textgreater{}=0.05)
\item [-{-}annotationExists=False]  
Optional. Select only the annotated variants
\item [-{-}annotationDoesNotExist=False]  
Optional. Select only the variants without annotation
\item [-{-}genotype]  
Optional. Select by sample genotype. Samples names must be specified as they are stored the db. Please, find more information in the documentation, Syntax: \{samp\_1\}:\{gt\_1\}(,\{gt\_n\})*(;\{samp\_n\}:\{gt\_1\}(,\{gt\_n\})*)* (e.g. HG0097:0/0;HG0098:0/1,1/1)
\item [-{-}consequence\_type]  
Optional. Select by consequence type. Consequence type SO term list. Use comma to separate the SO terms. Please, find the information of SO terms supported in the documentation. (e.g. SO:0000045,SO:0000046
\item [-{-}xref]  
Optional. Select by XRef, this is a field used to map ids from different dbs. Please, find the information of dbs supported in the documentation. 
\item [-{-}biotype]  
Optional. Select by Biotype. Consequence type SO term list. Use comma to separate the biotypes. Please, find the information of biotype terms supported in the documentation. (e.g. protein\_coding,retained\_intron. This filter should not be used alone due to the huge amount of results could be retrieved
\item [-{-}polyphen]  
Optional. Select by polyphen, polyphen score ranges from {[}0-1{]}. Syntax: {[}\textless{}\textbar{}\textgreater{}\textbar{}\textless{}=\textbar{}\textgreater{}={]}\{number\}, (e.g, \textgreater{}=0.9). This filter is slow if it is not used along others
\item [-{-}sift]  
Optional. Select by sift, sift score ranges from {[}0-1{]}. Syntax: {[}\textless{}\textbar{}\textgreater{}\textbar{}\textless{}=\textbar{}\textgreater{}={]}\{number\}, (e.g, \textgreater{}=0.9). This filter is slow if it is not used along others
\item [-{-}conservation]  
Optional. Select by conservation sources. Please read the documentation to find the sources available. Use comma to separate the conservation sources. Syntax: sourceName{[}\textless{}\textbar{}\textgreater{}\textbar{}\textless{}=\textbar{}\textgreater{}={]}number (e.g. phastCons\textgreater{}0.5,phylop\textless{}0.1). . This filter is slow if it is not used along others
\item [-{-}alternate\_frequency]  
Optional. Select by frequency of the alternate allele in one population. Please read the documentation to find the population available. Use comma to separate the population. Syntax: populationName{[}\textless{}\textbar{}\textgreater{}\textbar{}\textless{}=\textbar{}\textgreater{}={]}number (e.g. 1000g\_CEU\textgreater{}0.5,1000g\_AFR\textless{}0.1)
\item [-{-}reference\_frequency]  
Optional. Select by frequency of the reference allele in one population. Please read the documentation to find the population available. Use comma to separate the population. Syntax: populationName{[}\textless{}\textbar{}\textgreater{}\textbar{}\textless{}=\textbar{}\textgreater{}={]}number (e.g. 1000g\_CEU\textgreater{}0.5,1000g\_AFR\textless{}0.1)
\item [-{-}limit]  
Optional. limit (number of results)
\item [-{-}skip]  
Optional. skip (number of results)
\item [-{-}sort=False]  
Optional. Sort the output by chromosome coordinates
\item [-{-}group\_by]  
Optional. Group the output by 
\item [-{-}unknownGenotype=./.]  
Optional. Returned genotype for unknown genotypes. 
\item [-{-}returnedSamples]  
Optional. Only the specified samples will be returned. The samples names in the db, if you have doubts about this, please read the documentation. Use comma to separate the sample names
\item [-{-}returnedFiles]  
Optional. Only the information from the specified files will be returned.Please note these are the id files in the DB, use pyCGA to get the id files. Use comma to separate the files id
\item [-{-}batchSize=5000]  
Optional. This parameter control the size of the batches of variants per query, This number is proportional to the memory used and inversely proportional to the time. This value can not be greater than 5000.
\item [-{-}outputType=json]  
Optional. This parameter control the size of the batches of variants per query, This number is proportional to the memory used and inversely proportional to the time. This value can not be greater than 5000.

Possible choices: json, VCF, AVRO-GA4GH, AVRO-OPENCGA
\end{optionlist}

\end{description}


\subsection{Examples}
\label{pyVariantFetcher:examples}
Get all variants in chromosome 1 with a maf in the database less than 0.05:

\begin{Verbatim}[commandchars=\\\{\}]
pyCGAVariantFetcher [sid] \PYGZhy{}\PYGZhy{}host [hostname] \PYGZhy{}\PYGZhy{}studyID [studyID] \PYGZhy{}\PYGZhy{}chromosome 1 \PYGZhy{}\PYGZhy{}maf \PYGZlt{} 0.05
\end{Verbatim}

Get all nonsense variants in a certain gene panel with a phylop conservation score more than 0.2:

\begin{Verbatim}[commandchars=\\\{\}]
panel=PAX6,B3GALTL,SOX2,MFRP,RAX,BCOR,OTX2,SIX6,BMP4,ALDH1A3,COL4A1,BMPR1A,HCCS,CYP1B1,RAB3GAP1,SHH,VSX2,FOXC1,VAX1,FOXE3,FRAS1,FREM1,XPA,HDAC6,STRA6,FREM2,ERCC3,SMOC1,GRIP1,XPC,RAB18,ERCC2,DDB1,RAB3GAP2,DDB2,ABCB6,ERCC6,ERCC4,ERCC5,PITX2,POLH,GDF3,PITX3,ERCC8,GTF2H5,GDF6,ERCC1,MPLKIP,PRSS56,RARB,TENM3,MAB21L2,TBC1D20
pyCGAVariantFetcher [sid] \PYGZhy{}\PYGZhy{}host [hostname] \PYGZhy{}\PYGZhy{}studyID [studyID] \PYGZhy{}\PYGZhy{}consequenceType SO:1000062 \PYGZhy{}\PYGZhy{}gene \PYGZdl{}panel \PYGZhy{}\PYGZhy{}conservation phylop\PYGZgt{}0.2
\end{Verbatim}

Get all variants in a certain region with an alternate frequency in AMR population for 1000 Genomes less than 0.1:

\begin{Verbatim}[commandchars=\\\{\}]
pyCGAVariantFetcher [sid] \PYGZhy{}\PYGZhy{}host [hostname] \PYGZhy{}\PYGZhy{}studyID [studyID] \PYGZhy{}\PYGZhy{}region 1:1100000\PYGZhy{}1300000 \PYGZhy{}\PYGZhy{}alternate\PYGZus{}frequency
\end{Verbatim}

\begin{notice}{note}{Note:}
If you have doubts of how to get and use the ids in OpenCGA, pleas read {\hyperref[ids:how-to-use-ids]{\emph{\DUspan{}{Id Considerations in OpenCGA}}}}
\end{notice}

\begin{notice}{note}{Note:}
Please notice this program is fetching variants from a DB which can contain millions of variants.
Although the DB is indexed to obtain the best performance, several general queries are not supported to be
fast. In this way, for example,  if you try to fetch variants filtering only by type of variants (i,e SNVs),
the result will be very slow, but if you query for a type of variant in a specific region will be very fast.
\begin{description}
\item[{So we recommend, use at least one of these filters in your queries:}] \leavevmode\begin{itemize}
\item {} 
ids

\item {} 
chromosome

\item {} 
region

\item {} 
gene

\item {} 
genotype

\item {} 
consequenceType

\item {} 
xref

\item {} 
alternate\_frequency

\item {} 
reference\_frequency

\item {} 
maf

\item {} 
mgf

\end{itemize}

\end{description}
\end{notice}


\chapter{Developer Documentation}
\label{documentation:dev-docs}\label{documentation:developer-documentation}

\section{How to implement new web services parsers}
\label{new_web_services::doc}\label{new_web_services:how-to-implement-new-web-services-parsers}
inf



\renewcommand{\indexname}{Index}
\printindex
\end{document}
